### DISCUSSION TITLE
useFetch/useAsyncData will not fetch data on page refresh/reload (but works fine during navigation)

### DISCUSSION THREAD

**Discussion by ymansurozer** [nuxt/nuxt#15941] [Questions] [answered]
I am fetching data with a top level `useFetch` as below:

```
<script setup lang="ts">
...
const { data } = useFetch("/api/data");
...
</script>
```

This works fine when I navigate between routes. But when I refresh/reload the page from the browser or just copy and paste the url to a new tab, `data` is not fetched and is null. To make this work, I had to do the fetching manually `onMounted` like below:

```
<script setup lang="ts">
...
const { data, execute } = useFetch("/api/data");

onMounted(() => {
  if(!data.value) execute();
});
...
</script>
```

My question is: Is this the intended behavior? Why does useFetch not work when there is a page refresh/reload and is there another way to make it work (aside using an `onMounted` hook)?

**Comment by ymansurozer** [comment]
Now I also noticed an issue with my workaround of using `onMounted`. When I do that, nuxt/supabase module sometimes fails to send the authorization headers. So it is really a problem.

Does anyone has any idea why `useFetch` does not run on page refresh?

**Comment by wobsoriano** [comment]
You have to ‘await’ it.

**Comment by ymansurozer** [comment]
Sorry, that was a typo. I am doing await but still the same result. But I now realized this is related to the nuxt/supabase module because it does not send authorization headers to backend in top-level useFetch statements. Will move this issue there now.

**Reply by bayramorhan** [reply]
I think it has nothing to do with supabase. It s just an issue about useAsyncData function's behavior

**Comment by BlackSkyJY** [comment]
I also have the above problems. I did not create the server directory，I won't have the above problems if I use $fetch

**Comment by bayramorhan** [comment]
I also having similar problem. When I load the page first time it works perfectly but when I reload page, there is literally nothing. just null.

```
const { data, error} = await useAsyncData(
            () =>
              $fetch("/api/my-endpoint", {
                headers: { "context-token": token },
              })
          );
```

**Comment by ymansurozer** [comment, marked as answer]
Sorry people, I did not have the chance to add my conclusion from this issue. It seems to be more related to the nuxt/supabase module. Here is the discussion: https://github.com/nuxt-modules/supabase/issues/135

So apparently, we need to send the headers ourselves in frontend, otherwise they are not sent (for my case, sometimes).

**Reply by bayramorhan** [reply]
Im having the same issue but Im not using supabase.

**Reply by ymansurozer** [reply]
That's interesting. It works in my case, the only issue is it does not send the authorization headers to backend. 

I assume you've tried checking if the backend receives the request? I am asking because I first thought the issue was that request was not sent to the backend at all but I later realized the request was sent but only without auth details so my function was failing.

**Reply by bayramorhan** [reply]
Im using nuxt 3 server module. I confirmed request was not sent to backend.

**/server/api/my-api.js**
```

import axios from "axios";

export default defineEventHandler(async(event) => {
    console.log('Received new request') **//There is no log in the console**
    const contextToken = event.node.req.headers['context-token']
    const runtimeConfig = useRuntimeConfig();
    let cartData = null;
    let cartError = null;
    await axios.post(`${runtimeConfig.public.apiBase}/checkout/cart`, {}, { headers: { 'sw-access-key': runtimeConfig.apiKey, 'sw-context-token': contextToken } })
    .then((response) => {
        cartData = response.data;
    }).catch((err) => {
        cartError = err;
    })


    if(cartData) {
        event.node.res.statusCode = 200;
        return cartData
    } else if(cartError) {
        let error = cartError.response.data.errors[0];
        event.node.res.statusCode = error.status;
        event.node.res.statusMessage = error.detail;
        throw createError({ statusCode: error.status, message: error.detail, name: error.title })
    }

})
```

**Reply by bayramorhan** [reply]
maybe nuxt caching the request somehow and dont send it again on page reload.

**Comment by ymansurozer** [comment]
And you're using a top-level 'useFetch' to send the request to this
endpoint, right? It is weird. As far as I know, caching was removed from
the stable release of Nuxt, see here
https://github.com/nuxt/framework/issues/8917.

**Comment by FatNerdPeng** [comment]
I have same problem:
https://github.com/nuxt/nuxt/discussions/18726
I don't know why and how to solve it

**Comment by watzon** [comment]
Was having the same problem. Setting `server` to false in the options fixed it, but I don't consider this a real fix.

**Reply by carlosesquivelroisenvit** [reply]
I'm having the same issue, on a newly created app (Nuxt 3.13) I have a `await useAsyncData...` on the `<script setup>` of the layout and the data returned is always null. I have navigate or change any state in the page, then it's loaded correctly.

**Comment by guitarzero** [comment]
Hi all. 
Any news on this (except from supabase answer) ? It is quite annoying and i do want to have SSR so "server:false" is not an option. 
I also suspect a cache issue. 
Thank you

**Comment by vanghoa** [comment]
Hey, I encountered the same issue when using useAsyncData for dynamic routes in a static site built with npx nuxt generate, and I only want the fetching takes place during the build time. After some troubleshooting, I found a fix that works for me:

```
const { data, refresh } = await useAsyncData(
    slug,
    async () => {
        // Fetch data and return
    },
    { server: true, lazy: false }
);

if (import.meta.browser && !data.value) {
    await refresh();
}

```
This solution ensures that when running in the browser, the data gets refreshed if it's initially null (on hard reload or direct navigation from a url). Note that I'm checking !data.value before refreshing—without this check, refreshing could unintentionally set the data to null (during the SPA navigation with NuxtLink).

hope it helps

**Comment by gillesserrien** [comment]
Has anyone still encountered this issue?
